pkg/chunk/gcp/storage_client.go

```diff
@@ -0,0 +1,227 @@
+package gcp
+
+import (
+	"flag"
+	"strings"
+
+	"cloud.google.com/go/bigtable"
+	log "github.com/Sirupsen/logrus"
+	"golang.org/x/net/context"
+
+	"github.com/weaveworks/cortex/pkg/chunk"
+)
+
+const (
+	columnFamily = "f"
+	column       = "c"
+	separator    = "\000"
+)
+
+// Config for a StorageClient
+type Config struct {
+	project  string
+	instance string
+}
+
+// RegisterFlags adds the flags required to config this to the given FlagSet
+func (cfg *Config) RegisterFlags(f *flag.FlagSet) {
+	f.StringVar(&cfg.project, "bigtable.project", "", "BigTable project ID.")
+	f.StringVar(&cfg.instance, "bigtable.instance", "", "BigTable instance ID.")
+}
+
+// storageClient implements chunk.storageClient for GCP.
+type storageClient struct {
+	cfg       Config
+	schemaCfg chunk.SchemaConfig
+	client    *bigtable.Client
+}
+
+// NewStorageClient returns a new StorageClient.
+func NewStorageClient(ctx context.Context, cfg Config, schemaCfg chunk.SchemaConfig) (chunk.StorageClient, error) {
+	client, err := bigtable.NewClient(ctx, cfg.project, cfg.instance, instrumentation())
+	if err != nil {
+		return nil, err
+	}
+	return &storageClient{
+		cfg:       cfg,
+		schemaCfg: schemaCfg,
+		client:    client,
+	}, nil
+}
+
+func (s *storageClient) NewWriteBatch() chunk.WriteBatch {
+	return bigtableWriteBatch{
+		tables: map[string]map[string]*bigtable.Mutation{},
+	}
+}
+
+type bigtableWriteBatch struct {
+	tables map[string]map[string]*bigtable.Mutation
+}
+
+func (b bigtableWriteBatch) Add(tableName, hashValue string, rangeValue []byte, value []byte) {
+	rows, ok := b.tables[tableName]
+	if !ok {
+		rows = map[string]*bigtable.Mutation{}
+		b.tables[tableName] = rows
+	}
+
+	// TODO the hashValue should actually be hashed.
```

Comment:
Its also not clear this will affect load balancing on BigTable, as it dynamically splits ranges (completely different to how DynamoDB load balances data).